client/consensus/proposer/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Cumulus [`Proposer`] is a wrapper around a Substrate [`sp_consensus::Environment`]
+//! for creating
+//!
+//! This utility is designed to be composed within any collator consensus algorithm.
+
+use async_trait::async_trait;
+
+use sp_inherents::InherentData;
+use sp_consensus::{EnableProofRecording, Proposer as SubstrateProposer, Proposal, Environment};
+use sp_runtime::{traits::Block as BlockT, Digest};
+use sp_state_machine::StorageProof;
+use cumulus_primitives_parachain_inherent::ParachainInherentData;
+
+use std::error::Error as StdError;
+use std::fmt::Debug;
+use std::time::Duration;
+
+/// Errors that can occur when proposing a parachain block.
+pub struct Error {
+	inner: ErrorInner,
+}
+
+impl std::fmt::Display for Error {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self.inner {
+			ErrorInner::ProposerCreation(_) => write!(f, "Unable to create block proposer"),
+			ErrorInner::Proposing(_) => write!(f, "Unable to propose block"),
+		}
+	}
+}
+
+impl Debug for Error {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self.inner {
+			ErrorInner::ProposerCreation(ref e) => write!(f, "Unable to create block proposer: {:?}", e),
+			ErrorInner::Proposing(ref e) => write!(f, "Unable to propose block: {:?}", e),
+		}
+	}
+}
+
+impl StdError for Error {}
+
+impl Error {
+	fn proposer_creation(x: impl StdError + 'static) -> Self {
+		Error {
+			inner: ErrorInner::ProposerCreation(Box::new(x)),
+		}
+	}
+
+	fn proposing(x: impl StdError + 'static) -> Self {
+		Error {
+			inner: ErrorInner::Proposing(Box::new(x)),
+		}
+	}
+}
+
+#[derive(Debug)]
+enum ErrorInner {
+	ProposerCreation(Box<dyn StdError>),
+	Proposing(Box<dyn StdError>),
+}
+
+/// An interface for proposers.
+#[async_trait]
+pub trait ProposerInterface<Block: BlockT> {
+	/// The underlying DB transaction type produced with the block proposal.
+	type Transaction: Default + Send + 'static;
+
+	/// Propose a collation using the supplied `InherentData` and the provided
+	/// `ParachainInherentData`.
+	///
+	/// Also specify any required inherent digests, the maximum proposal duration,
+	/// and the block size limit in bytes. See the documentation on [`sp_consensus::Proposer::propose`]
+	/// for more details on how to interpret these parameters.
+	///
+	/// The `InherentData` and `Digest` are left deliberately general in order to accommodate
+	/// all possible collator selection algorithms or inherent creation mechanisms,
+	/// while the `ParachainInherentData` is made explicit so it will be constructed appropriately.
+	///
+	/// If the `InherentData` passed into this function already has a `ParachainInherentData`,
+	/// this will throw an error.
```

Comment:
How's it guaranteed?

Reply:
At the implementation level - replaced 'will' with 'should'.